src/v/kafka/server/tests/fetch_bench.cc

```diff
@@ -0,0 +1,202 @@
+#include "model/fundamental.h"
+#include "model/namespace.h"
+#include "redpanda/tests/fixture.h"
+#include "test_utils/async.h"
+#include "test_utils/fixture.h"
+#include "test_utils/scoped_config.h"
+
+#include <seastar/testing/perf_tests.hh>
+#include <seastar/testing/thread_test_case.hh>
+
+#include <boost/test/tools/old/interface.hpp>
+
+ss::logger flog("fetch_bench");
+
+struct fetch_bench_fixture : redpanda_thread_fixture {
+    // Topic settings
+    static constexpr size_t topic_name_length = 30;
+
+    // Message settings
+    static constexpr size_t batch_size = 1024;
+    static constexpr size_t batch_overhead = 70;
+
+    struct fetch_part {
+        model::partition_id pid;
+        model::offset offset;
+        size_t num_batches;
+    };
+
+    struct fetch_topic {
+        model::topic name;
+        std::vector<fetch_part> partitions;
+    };
+
+    ss::future<> fetch_from(model::topic t, std::vector<fetch_part> parts) {
+        auto conn_context = make_connection_context();
+        co_await conn_context->start();
+
+        // make the fetch topic
+        kafka::fetch_topic ft;
+        ft.name = t;
+
+        size_t total_batches = 0;
+        // add the partitions to the fetch request
+        for (const auto& part : parts) {
+            kafka::fetch_partition fp;
+            fp.partition_index = model::partition_id(part.pid);
+            fp.fetch_offset = part.offset;
+            fp.current_leader_epoch = kafka::leader_epoch(-1);
+            fp.log_start_offset = model::offset(-1);
+            fp.max_bytes = part.num_batches * (batch_size + batch_overhead);
+            ft.fetch_partitions.push_back(std::move(fp));
+            total_batches += part.num_batches;
+        }
+
+        // create a request
+        kafka::fetch_request_data frq_data;
+        frq_data.replica_id = kafka::client::consumer_replica_id;
+        frq_data.max_wait_ms = 500ms;
+        frq_data.min_bytes = total_batches * (batch_size + batch_overhead);
+        frq_data.max_bytes = 1048576;
+        frq_data.isolation_level = model::isolation_level::read_uncommitted;
+        frq_data.session_id = kafka::invalid_fetch_session_id;
+        frq_data.session_epoch = kafka::final_fetch_session_epoch;
+        frq_data.topics.push_back(std::move(ft));
+
+        auto request = kafka::fetch_request{std::move(frq_data)};
+
+        kafka::request_header header{
+          .key = kafka::fetch_handler::api::key,
+          .version = kafka::fetch_handler::max_supported};
+
+        auto rctx = make_request_context(
+          std::move(request), header, conn_context);
+        auto octx = kafka::op_context(
+          std::move(rctx), ss::default_smp_service_group());
+
+        // Drain task queue before running the measured region in order to
+        // reduce noise from unrelated tasks.
+        co_await tests::drain_task_queue();
+
+        perf_tests::start_measuring_time();
+        co_await kafka::testing::do_fetch(
```

Comment:
Would it make sense to fetch N times here (with limited size) or do we think that's a different benchmark altogether?

Reply:
Switched to fetch 100 times here to reduce the impact of timer overhead on the overall results.